DisCatSharp.ApplicationCommands/Attributes/ContextMenu/ContextMenuCooldownAttribute.cs

```diff
@@ -0,0 +1,305 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+using System;
+using System.Collections.Concurrent;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+
+using DisCatSharp.ApplicationCommands.Context;
+using DisCatSharp.ApplicationCommands.Enums;
+
+namespace DisCatSharp.ApplicationCommands.Attributes;
+
+/// <summary>
+/// Defines a cooldown for this command. This allows you to define how many times can users execute a specific command
+/// </summary>
+[AttributeUsage(AttributeTargets.Method | AttributeTargets.Class, AllowMultiple = true, Inherited = false)]
+public sealed class ContextMenuCooldownAttribute : ContextMenuCheckBaseAttribute
+{
+	/// <summary>
+	/// Gets the maximum number of uses before this command triggers a cooldown for its bucket.
+	/// </summary>
+	public int MaxUses { get; }
+
+	/// <summary>
+	/// Gets the time after which the cooldown is reset.
+	/// </summary>
+	public TimeSpan Reset { get; }
+
+	/// <summary>
+	/// Gets the type of the cooldown bucket. This determines how cooldowns are applied.
+	/// </summary>
+	public CooldownBucketType BucketType { get; }
+
+	/// <summary>
+	/// Gets the cooldown buckets for this command.
+	/// </summary>
+	private readonly ConcurrentDictionary<string, ContextMenuCooldownBucket> _buckets;
+
+	/// <summary>
+	/// Defines a cooldown for this command. This means that users will be able to use the command a specific number of times before they have to wait to use it again.
+	/// </summary>
+	/// <param name="maxUses">Number of times the command can be used before triggering a cooldown.</param>
+	/// <param name="resetAfter">Number of seconds after which the cooldown is reset.</param>
+	/// <param name="bucketType">Type of cooldown bucket. This allows controlling whether the bucket will be cooled down per user, guild, channel, or globally.</param>
+	public ContextMenuCooldownAttribute(int maxUses, double resetAfter, CooldownBucketType bucketType)
+	{
+		this.MaxUses = maxUses;
+		this.Reset = TimeSpan.FromSeconds(resetAfter);
+		this.BucketType = bucketType;
+		this._buckets = new ConcurrentDictionary<string, ContextMenuCooldownBucket>();
+	}
+
+	/// <summary>
+	/// Gets a cooldown bucket for given command context.
+	/// </summary>
+	/// <param name="ctx">Command context to get cooldown bucket for.</param>
+	/// <returns>Requested cooldown bucket, or null if one wasn't present.</returns>
+	public ContextMenuCooldownBucket GetBucket(ContextMenuContext ctx)
+	{
+		var bid = this.GetBucketId(ctx, out _, out _, out _);
```

Comment:
This looks weird. If I just want the bucket ID I'm *required* to discard 3 things.

Reply:
Do you have another idea?

Reply:
After looking at it again, this is a private method only used in 1 place. If the 1 time we call this method discards those output vars, we really don't need it? Unless there's more coming that hasn't been unveiled yet